### ISSUE TITLE
runtime timers are unimplemented

### ISSUE THREAD

**Issue by ryskiwt** [tinygo-org/tinygo#1037]
Using time.NewTicker causes a runtime error:

```go
package main

import (
	"time"
)

func main() {
	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()

	<-ticker.C

	neverDone := make(chan struct{})
	<-neverDone
}
```

Error:

```
Uncaught (in promise) LinkError: WebAssembly.instantiate(): Import #12 module="env" function="time.startTimer" error: function import requires a callable
```

Compiled using the 0.12.0 docker image:

```
docker run ... \
    -e GOOS=js -e GOARCH=wasm \
    tinygo/tinygo:0.12.0 tinygo build -target wasm -o hoge.wasm .
```

**Comment by jaddr2line** [comment]
TinyGo does not implement the `runtimeTimer` system which is used to implement `time.NewTicker` and `time.AfterFunc`. We either need to somehow implement it (not currently possible due to how `//go:linkname` is implemented), or replace parts of the `time` package.

**Comment by ryskiwt** [comment]
Thanks for reply. I got it.
Is there any timeline or roadmap for implementing `go:linkname` ?

**Comment by jaddr2line** [comment]
So `//go:linkname` has been implemented, but in TinyGo it preserves type information. Since `time.runtimeTimer` is an unexported type, another package cannot linkname `time.startTimer`, because the types would be different and cause a compiler error.

**Comment by jaddr2line** [comment]
For now, if possible, use `time.Sleep` or loops. I do not know when we are going to be able to efficiently implement this.

**Comment by jaddr2line** [comment]
As a workaround, I wonder whether we could do something like `//go:linkname` but with types.

**Comment by aykevl** [comment]
Maybe we need to support `//go:linkname` with slightly-mismatched types after all. Yes, we could allow custom type names for named types, but doing something like that would still be quite invasive. I still don't like the `//go:extern` trick for globals, it requires parsing the AST. I would prefer to avoid one more dependency on having the AST available while building the LLVM IR.

Slightly-mismatched types would be much less of a problem if [opaque pointer types](https://groups.google.com/forum/#!topic/llvm-dev/Dw_DYSXGFto) were completed in LLVM. Unfortunately, it looks like it'll still take quite a while (if it is ever finished). Until then, maybe the simplest solution is the following: if a function to call has a changed linkname (with `//go:linkname`), then check whether the types are the same and if not bitcast the function pointer. I think that would be the least invasive change to the compiler. Perhaps it would be easier after #1008.

**Comment by deadprogram** [comment]
I was just doing some experimenting with trying to run https://github.com/mbertschler/dragon-iss-docking-autopilot using TinyGo and ran into this exact problem pretty quickly. 

Was reading the proposed work-arounds. Any more ideas on this currently, or do we just need to get #1008 merged in and proceed from there?

**Comment by aykevl** [comment]
#1008 is not really related, it would just make one implementation a bit easier.

I had another idea how `//go:linkname` could be fixed: every pointer to a named struct could be cast to `*i8` or similar during the call. I don't expect that to affect optimizations, but it should be checked to be sure. It does add a bit of noise to the IR, however. And it might make some TinyGo optimization passes more complicated.

**Comment by deadprogram** [comment]
I still really would like to be able to write code like 

```
case <-time.NewTimer(3 * time.Second).C:
```

**Comment by aykevl** [comment]
Yes, and we will need this to support the stdlib testing package (which makes use of these timers).

**Comment by aykevl** [comment]
I have an implementation for timers and tickers, will clean up the patch and send a PR soon.

**Comment by aykevl** [comment]
Here is the PR: #1402

**Comment by Zyko0** [comment]
Any update on this ?

**Comment by aykevl** [comment]
I think I can take a look after #2011 has been merged. The current PR is way out of date.

**Comment by Zyko0** [comment]
Any idea about when this could happen now #2011 has been merged ?
Thanks !

**Comment by deadprogram** [comment]
Released as part of `v0.26.0`
